// @aws-solutions-constructs/resources/lib/key-policy-updater.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKeyPolicyUpdaterCustomResource = createKeyPolicyUpdaterCustomResource;
const iam = require("aws-cdk-lib/aws-iam");
const lambda = require("aws-cdk-lib/aws-lambda");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const custom_resources_1 = require("aws-cdk-lib/custom-resources");
const core_1 = require("@aws-solutions-constructs/core");
const utils_1 = require("./utils");
const defaults = require("@aws-solutions-constructs/core");
function createKeyPolicyUpdaterCustomResource(scope, id, props) {
    const lambdaFunction = (0, core_1.buildLambdaFunction)(scope, {
        lambdaFunctionProps: {
            runtime: defaults.COMMERCIAL_REGION_LAMBDA_NODE_RUNTIME,
            handler: 'index.handler',
            description: 'Custom resource function that updates a provided key policy to allow CloudFront access.',
            timeout: props.timeout,
            memorySize: props.memorySize,
            code: lambda.Code.fromAsset(`${__dirname}/key-policy-updater-custom-resource`)
        }
    });
    const customResourceCmkPrivilegePolicy = new iam.Policy(scope, `${id}ResourceCmkPolicy`, {
        statements: [
            new iam.PolicyStatement({
                actions: ['kms:PutKeyPolicy', 'kms:GetKeyPolicy', 'kms:DescribeKey'],
                effect: iam.Effect.ALLOW,
                resources: [props.encryptionKey.keyArn]
            })
        ]
    });
    lambdaFunction.role?.attachInlinePolicy(customResourceCmkPrivilegePolicy);
    const kmsKeyPolicyUpdateProvider = new custom_resources_1.Provider(scope, 'KmsKeyPolicyUpdateProvider', {
        onEventHandler: lambdaFunction
    });
    (0, utils_1.addCfnSuppressRulesForCustomResourceProvider)(kmsKeyPolicyUpdateProvider);
    const customResource = new aws_cdk_lib_1.CustomResource(scope, 'KmsKeyPolicyUpdater', {
        resourceType: 'Custom::KmsKeyPolicyUpdater',
        serviceToken: kmsKeyPolicyUpdateProvider.serviceToken,
        properties: {
            KmsKeyId: props.encryptionKey.keyId,
            CloudFrontDistributionId: props.distribution.distributionId,
            AccountId: aws_cdk_lib_1.Aws.ACCOUNT_ID
        },
    });
    return {
        lambdaFunction,
        customResource
    };
}
