// @aws-solutions-constructs/aws-cloudfront-s3/test/integ.cfts3-no-cloudfront-s3-access-logs.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
// Imports
const aws_cdk_lib_1 = require("aws-cdk-lib");
const lib_1 = require("../lib");
const core_1 = require("@aws-solutions-constructs/core");
const integ_tests_alpha_1 = require("@aws-cdk/integ-tests-alpha");
const defaults = require("@aws-solutions-constructs/core");
// Setup
const app = new aws_cdk_lib_1.App();
const stack = new aws_cdk_lib_1.Stack(app, (0, core_1.generateIntegStackName)(__filename));
stack.templateOptions.description = 'Integration Test for aws-cloudfront-s3';
const construct = new lib_1.CloudFrontToS3(stack, 'test-cloudfront-s3', {
    logS3AccessLogs: false,
    logCloudFrontAccessLog: false,
    cloudFrontDistributionProps: {},
    cloudFrontLoggingBucketProps: {
        removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
        autoDeleteObjects: true
    },
    bucketProps: {
        removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
        autoDeleteObjects: true
    }
});
defaults.addCfnSuppressRules(construct.cloudFrontWebDistribution, [
    { id: 'W10',
        reason: 'Test case only' },
]);
defaults.addCfnSuppressRules(construct.s3Bucket, [
    { id: 'W35',
        reason: 'Test case only' },
]);
defaults.addCfnSuppressRules(construct.cloudFrontLoggingBucket, [
    { id: 'W35',
        reason: 'Test case only' },
]);
defaults.addCfnSuppressRules(construct.cloudFrontWebDistribution, [
    { id: 'W35',
        reason: 'Test case only' },
]);
defaults.addCfnGuardSuppressRules(construct.cloudFrontLoggingBucket, ["S3_BUCKET_LOGGING_ENABLED"]);
defaults.suppressCustomHandlerCfnNagWarnings(stack, 'Custom::S3AutoDeleteObjectsCustomResourceProvider');
// Synth
new integ_tests_alpha_1.IntegTest(stack, 'Integ', { testCases: [
        stack
    ] });
