// upsert-slr/lib/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ServiceLinkedRole = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const fs_1 = require("fs");
const path_1 = require("path");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const constructs_1 = require("constructs");
class ServiceLinkedRole extends constructs_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const handler = new aws_lambda_1.SingletonFunction(this, 'CustomResourceHandler', {
            // Use raw string to avoid from tightening CDK version requirement
            runtime: new aws_lambda_1.Runtime('nodejs20.x', aws_lambda_1.RuntimeFamily.NODEJS, { supportsInlineCode: true }),
            code: aws_lambda_1.Code.fromInline((0, fs_1.readFileSync)((0, path_1.join)(__dirname, '../', 'lambda', 'dist', 'index.js')).toString()),
            handler: 'index.handler',
            uuid: '8f7be66a-3315-474b-aea1-6ceca43d27c3', // generated for this construct
            lambdaPurpose: 'UpsertSlrCustomResourceHandler',
            timeout: aws_cdk_lib_1.Duration.minutes(3),
            memorySize: 128,
        });
        handler.addToRolePolicy(new aws_iam_1.PolicyStatement({
            actions: ['iam:CreateServiceLinkedRole'],
            resources: ['*'],
        }));
        const properties = {
            awsServiceName: props.awsServiceName,
            description: props.description,
            waitTimeSeconds: aws_cdk_lib_1.Duration.minutes(1).toSeconds(),
        };
        new aws_cdk_lib_1.CustomResource(this, 'Default', {
            serviceToken: handler.functionArn,
            resourceType: 'Custom::UpsertServiceLinkedRole',
            properties,
        });
    }
}
exports.ServiceLinkedRole = ServiceLinkedRole;
_a = JSII_RTTI_SYMBOL_1;
ServiceLinkedRole[_a] = { fqn: "upsert-slr.ServiceLinkedRole", version: "1.0.6" };
