// @aws-solutions-constructs/resources/test/integ.key-policy-updater.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const aws_cdk_lib_1 = require("aws-cdk-lib");
const core_1 = require("@aws-solutions-constructs/core");
const integ_tests_alpha_1 = require("@aws-cdk/integ-tests-alpha");
const s3 = require("aws-cdk-lib/aws-s3");
const kms = require("aws-cdk-lib/aws-kms");
const cft = require("aws-cdk-lib/aws-cloudfront");
const cfto = require("aws-cdk-lib/aws-cloudfront-origins");
const key_policy_updater_1 = require("../lib/key-policy-updater");
const utils_1 = require("../../core/lib/utils");
const app = new aws_cdk_lib_1.App();
const stack = new aws_cdk_lib_1.Stack(app, (0, core_1.generateIntegStackName)(__filename));
stack.templateOptions.description = 'Integration Test for Key Policy Updater Resource';
const key = new kms.Key(stack, 'test key', {
    removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
    enableKeyRotation: true
});
const bucket = new s3.Bucket(stack, 'test bucket', {
    removalPolicy: aws_cdk_lib_1.RemovalPolicy.DESTROY,
    encryption: s3.BucketEncryption.S3_MANAGED,
});
(0, utils_1.addCfnSuppressRules)(bucket, [{
        id: "W35",
        reason: "Test resource"
    }, {
        id: "41",
        reason: "Test resource"
    }
]);
const distribution = new cft.Distribution(stack, 'test dist', {
    defaultBehavior: {
        origin: new cfto.S3Origin(bucket),
    }
});
(0, utils_1.addCfnSuppressRules)(distribution, [{
        id: "W10",
        reason: "Test resource"
    }, {
        id: "W70",
        reason: "Test resource"
    }
]);
(0, key_policy_updater_1.createKeyPolicyUpdaterCustomResource)(stack, 'Test', {
    encryptionKey: key,
    distribution
});
new integ_tests_alpha_1.IntegTest(stack, 'Integ', {
    testCases: [
        stack
    ]
});
